upgrade/upgrade_storageclass.go

```diff
@@ -0,0 +1,100 @@
+package upgrade
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	storagev1 "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"github.com/longhorn/longhorn-manager/datastore"
+	"github.com/longhorn/longhorn-manager/types"
+)
+
+const (
+	envLonghornStorageClassSettingPath = "LONGHORN_STORAGECLASS_SETTING_PATH"
+)
+
+func doLonghornStorageClassUpgrade(namespace string, kubeClient *clientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrap(err, "error while upgrading Longhorn storageclass")
+	}()
+	sc, err := getLonghornStorageClassSetting()
+	if err != nil {
+		logrus.Warnf("Skip updating Longhorn storageclass. Cannot read the Longhorn storageclass from configmap : %v", err)
+		return nil
+	}
+
+	existingSC, err := kubeClient.StorageV1().StorageClasses().Get(types.LonghornStorageClassName, metav1.GetOptions{})
+	if err != nil {
+		logrus.Warn("Cannot get the existing Longhorn storageclass. Will overwrite it with new Longhorn storageclass")
+	}
+
+	if storageclassesHaveSameValues(existingSC, sc) {
+		logrus.Info("The Longhorn storageclass is already up to date")
+		return nil
+	}
+
+	err = kubeClient.StorageV1().StorageClasses().Delete(types.LonghornStorageClassName, nil)
+	if err != nil && !datastore.ErrorIsNotFound(err) {
+		logrus.Warnf("cannot delete the existing Longhorn storageclass. Skip updating Longhorn storageclass: %v", err)
+		return nil
+	}
+
+	_, err = kubeClient.StorageV1().StorageClasses().Create(sc)
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			logrus.Warnf("An up to date Longhorn storageclass already existed.")
+			return nil
+		}
+		return err
+	}
+
+	return nil
+}
+
+func getLonghornStorageClassSetting() (*storagev1.StorageClass, error) {
+	settingPath := os.Getenv(envLonghornStorageClassSettingPath)
+	if settingPath == "" {
+		return nil, fmt.Errorf("LonghornStorageClassSettingPath is empty")
+	}
+
+	data, err := ioutil.ReadFile(settingPath)
+
+	decode := scheme.Codecs.UniversalDeserializer().Decode
+	obj, _, err := decode(data, nil, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error while decoding YAML object")
+	}
+
+	sc, ok := obj.(*storagev1.StorageClass)
+	if !ok {
+		return nil, fmt.Errorf("%v is not a storageclass yaml", settingPath)
```

Comment:
What will happen if the storage yaml is invalid and result in upgrade failure? Will the manager restart automatically?

Reply:
No. The upgrade will not fail. It will skip the new storageclass YAML file and keep the old one. I consider updating `longhorn` storageclass as a best-effort action. If it fail, the users will likely to notice it and take some manual interventions. 
In this case, restart manager will not solve the problem because the next time it wakes up it will see the same invalid storageclass yaml file.

Reply:
I think that there is only 1 case that we need to stop the upgrade which is: we successfully deleted the existing `longhorn` storageclass but fail to create a new one. We need to stop the world in this case because if we don't workloads that are using `longhorn` storageclass will not be able to scale up.

Reply:
I don't follow. The current behavior is the manager going to restart, right? So that's upgrade failure.

What are we going to stop? All the managers would be done if the upgrade failed and user need to fix the config.

Reply:
> I don't follow. The current behavior is the manager going to restart, right? So that's upgrade failure.

I am sorry for the confusion. No, the current behavior is the manager will not restart. It skips upgrading storageclass and continue. The return error is swallowed here: https://github.com/longhorn/longhorn-manager/blob/926e7f9d023375629236e96f60092a9a41a360ae/upgrade/upgrade_storageclass.go#L30

> What are we going to stop when you said stop the world? All the managers would be done if the upgrade failed and user needs to fix the config.

By saying `stop the world`, I mean that when `a manager pod successfully deleted the existing longhorn storageclass but failed to create a new one.`, that manager will be restarted. The next time it run, it will try to recreate the storageclass.

Reply:
The problem of not stopping the manager is, the user will never notice the storage class error. We don't have anything to indicate the wrong configuration of storageclass except for the error in the log. And no user will look into the log unless something bad happens.

Another question is, what if the user changes the storageclass config map after the upgrade? Or does user suppose to change the storageclass config map after the upgrade? I am wondering should we keep the storageclass up to date to the configmap, not only for the upgrade process, but also during the normal operation.

Reply:
@yasker 

> The problem of not stopping the manager is, the user will never notice the storage class error. We don't have anything to indicate the wrong configuration of storageclass except for the error in the log. And no user will look into the log unless something bad happens.

I absolutely agree with this! I never look at the log unless something went wrong.  :)))) I fixed the code so that it will stop the manager incase the storageclass yaml file is invalid.


> Another question is, what if the user changes the storageclass config map after the upgrade? Or does user suppose to change the storageclass config map after the upgrade? I am wondering should we keep the storageclass up to date to the configmap, not only for the upgrade process, but also during the normal operation.

No, the users are not supposed to change the storageclass config map after the upgrade. It will confuse Helm and create the same problem as what we are having right now with the storageclass.

Reply:
Though any change to the configmap will be picked up by the manager upon restart, would that be a problem? ConfigMap should be changeable from helm's perspective? Since no field is prohibited from updating.

Reply:
I agree!